SARAALERT-1110: Vaccine Table

kylietmo opened the pull request:
# Description
Jira Ticket: [SARAALERT-1110](https://tracker.codev.mitre.org/browse/SARAALERT-1110)

This PR adds a new feature to the application for tracking vaccinations of a monitoree. It includes the following:
- New vaccinations table on the monitoree's page
- New Vaccine model, controller, and table in the database for tracking/handling vaccine records
- Appropriate handling of vaccine data (purging, filtering) 
- Tests for all of the above
**NOTE: There are two stub files (`app/javascript/tests/patient/vaccines/VaccineModal.test.js` and `app/javascript/tests/patient/vaccines/VaccineTable.test.js`) for the needed frontend tests. I will be filling those in before this is merged. Everything else, however, is ready for review.**

For some context, the available data options for vaccine data need to be based on CDC values for vaccine group and product name. We eventually will likely need to CVX codes for vaccines as well (again from the same source), as a result we decided to maintain the data we currently support in a configuration file that we will need to maintain and update accordingly. 

Support for export/import will be done in a separate PR.

# Demo
https://user-images.githubusercontent.com/11698457/108736477-a80f7580-74ff-11eb-86c4-f5b79264a05f.mov

# Important Changes
Please list important files (meaning substantial or integral to the PR) along with a list of the general changes that should be highlighted for reviewers.

`config/vaccines.yml`
- New vaccine configuration data.

`config/application.rb`
- Loading vaccines configuration file in as rails custom configuration. Access in Vaccine model.
 
`app/models/patient.rb`
`app/models/vaccine.rb`
- Vaccines belong to patients.
- Included validation for all fields on the vaccine model.

`db/migrate/20210204191950_create_vaccines.rb`
- Migration for new vaccine table.

`app/javascript/components/patient/vaccines/VaccineTable.js`
`app/javascript/components/patient/vaccines/VaccineModal.js`
- Frontend components for vaccines feature.

`app/helpers/vaccine_query_helper.rb`
`app/controllers/vaccines_controller.rb`
`config/routes.rb`
- Handles loading the vaccines table data as well as creating, and updating vaccine records. 

`app/views/patients/show.html.erb`
- Added the vaccine table frontend component to patient page.

`app/models/user.rb`
- User permissions relating to vaccines.

`app/jobs/purge_job.rb`
- Vaccines will be purged in the purge job.

`config/initializers/filter_parameter_logging.rb`
- Filtering vaccine data in logs.

`app/models/history.rb`
- Vaccine history items.

`app/javascript/packs/stylesheets/custom_table.scss`
`app/javascript/components/layout/CustomTable.js`
- Adding new `custom-table` class to handle vertical overflow in custom tables with dropdowns inside of them.

`app/javascript/components/assessment/AssessmentTable.js`
- Noticed a small inconsistency relating to modal bools that wasn't causing any actual issues but was confusing.

`lib/tasks/demo.rake`
- Generating demo dummy data.

# Testing
This fix was tested on the following browsers (submitter must check all those that apply):
* [x] Chrome
* [x] Firefox
* [x] Safari
* [ ] IE11

Please describe the tests needed to verify your changes. Provide instructions for reproducing these tests. List any relevant details for your test configuration.

It's best to test with more options in the vaccines configuration file. This is easily done by adding more dummy data to `vaccines.yml` like so:
```
'COVID-19':
  name: 'COVID-19'
  vaccines:
    - product_name: 'Moderna COVID-19 Vaccine'
    - product_name: 'Pfizer-BioNTech COVID-19 Vaccine'
'TestGroup':
  name: 'TestGroup'
  vaccines:
    - product_name: 'TestVaccine'
```

Additionally, please test the following (and anything else you can think of):
- Creating a vaccine record
- Editing a vaccine record
- Viewing vaccine records
- Trying to do all of the above as different users (only some users have vaccine permissions)
- Sorting vaccine columns
- Searching in the vaccines table
- Verify that the appropriate history items are created when vaccines are created/updated
- Running the purge job and verifying vaccine data is purged

kylietmo commented:
Working on addressing failing tests.

hackrm reviewed (commented):
Was able to spend some time testing out the UI aspects of the PR in all supported browsers. Here is the feedback and notes I have so far:

1. I see that the Jira ticket mentions an Administration Date 2 which is not included in this PR. Did the PO team decide to only have one Admin Date associated with a Vaccination record since you included the Dose Number field?
2. Should a selectable blank option be available for the Dose Number field? With the current behavior, once a user selects an option from the drop down, they can't clear it out and are forced to select a value from the drop down.
3. For the search box, for the most part it seems to work. But as soon as a user types in a search term that doesn't appear to return any vaccination records, the table does not display the "No data available in table" label and instead keeps displaying the set of records from the most recent successful search. For example, if you type in "COVID-19", all vaccinations with group "COVID-19" display as expected, but if you then search for "COVID-19test", the results displayed in the table will remain the same even though they no longer match the search term.
4. History entries seem to be created correctly, but the "Vaccination" and "Vaccination Edit" history type filters don't appear to work. When I filter for the "Vaccination" type history entries, no history entries are displayed even when I would expect to see some.
5. If the user is not on page 1 of the table and they search for a term, the table doesn't automatically set the table back to page 1.
6. I noticed some strange behavior with sorting columns when there are blank notes. I believe the expected behavior for sorting the "Notes" column is that all records with blank notes will always appear at the bottom of the sorted list regardless of what order the column is sorted in. But I found a way to break that behavior: (1) create a vaccination record with some text in the notes field (2) edit that new record and clear out the notes field so it is now blank and save it (3) sort the Notes column in the table in both ascending & descending order & observe the record you just updated with a blank notes field unexpectedly still displays in the top part of the sorted list with records that do have text in their Notes fields.
7. When updating the Number of Entries to display in the table, the warning in the screenshot below appears (same thing happens when updating the Number of Entries in the Symptom Assessments table).
![Screen Shot 2021-02-23 at 11 03 36 AM](https://user-images.githubusercontent.com/42656458/108878179-fab76300-75cd-11eb-9a90-497a2528bdb8.png)

kylietmo commented:
@hackrm 
> 1. I see that the Jira ticket mentions an Administration Date 2 which is not included in this PR. Did the PO team decide to only have one Admin Date associated with a Vaccination record since you included the Dose Number field?

Yes, it went through many iterations before a decision was made. I will update the ticket. 

> 2. Should a selectable blank option be available for the Dose Number field? With the current behavior, once a user selects an option from the drop down, they can't clear it out and are forced to select a value from the drop down.

Yes, there is a selectable blank option at the bottom.

For all your other points - great catches and I will investigate & address! Thank you!

hackrm commented on config/initializers/filter_parameter_logging.rb:
```diff
@@ -51,5 +51,8 @@
 Rails.application.config.filter_parameters += %i[comment]
+
+# Vaccine filters
+Rails.application.config.filter_parameters += %i[group_name product_name administration_date dose_number notes ]
```
Looks like there is an unnecessary space between `notes` and the ending bracket.

kylietmo replied:
Fixed here: 360c412

hackrm reviewed (commented):
Couple minor notes from my initial code review.

kylietmo commented on app/javascript/components/patient/vaccines/VaccineTable.js:
```diff
@@ -0,0 +1,422 @@
+   * Returns updated table data via an axios GET request.
+   * Debounces the query to avoid too many querys at once when someone is typing in the search bar, for example.
+   * @param {Object} query - Updated query for table data after change.
+   * @param {Boolean} isInitialLoad - Flag for if it's the initial load on component mount. Used to determine if colData needs to be updated.
```
Yes, good catch. Done here: 360c412

kylietmo commented:
@hackrm Fantastic catches - thank you! All have been fixed, see details below.

> 3. For the search box, for the most part it seems to work. But as soon as a user types in a search term that doesn't appear to return any vaccination records, the table does not display the "No data available in table" label and instead keeps displaying the set of records from the most recent successful search. For example, if you type in "COVID-19", all vaccinations with group "COVID-19" display as expected, but if you then search for "COVID-19test", the results displayed in the table will remain the same even though they no longer match the search term.

Ah, this was due to a bad check for the existence of the `total` value on the returned table data. If it was 0, nothing was being updated. Fixed here: [02fae9c](https://github.com/SaraAlert/SaraAlert/pull/721/commits/02fae9c2b9e1f2a60521a990d41d4d775ea0147d#diff-36bdccbc821e45bb9795ad5d5f73dbf2f7030b1837d7590d14dccb54d8cd3eb1R94)

> 4. History entries seem to be created correctly, but the "Vaccination" and "Vaccination Edit" history type filters don't appear to work. When I filter for the "Vaccination" type history entries, no history entries are displayed even when I would expect to see some.

This was because apparently the HistoryComponent expects the key names of the HISTORY_TYPES objects to match the displayed labels exactly. Fixed here: [02fae9c](https://github.com/SaraAlert/SaraAlert/pull/721/commits/02fae9c2b9e1f2a60521a990d41d4d775ea0147d#diff-08422754c6b565c51a7e094488199ec370790128736cd382aa486715d3126452R22)

> 5. If the user is not on page 1 of the table and they search for a term, the table doesn't automatically set the table back to page 1.

Just needed to do a page reset in this handler. Fixed here: [02fae9c](https://github.com/SaraAlert/SaraAlert/pull/721/commits/02fae9c2b9e1f2a60521a990d41d4d775ea0147d#diff-36bdccbc821e45bb9795ad5d5f73dbf2f7030b1837d7590d14dccb54d8cd3eb1R134)

> 6. I noticed some strange behavior with sorting columns when there are blank notes. I believe the expected behavior for sorting the "Notes" column is that all records with blank notes will always appear at the bottom of the sorted list regardless of what order the column is sorted in. But I found a way to break that behavior: (1) create a vaccination record with some text in the notes field (2) edit that new record and clear out the notes field so it is now blank and save it (3) sort the Notes column in the table in both ascending & descending order & observe the record you just updated with a blank notes field unexpectedly still displays in the top part of the sorted list with records that do have text in their Notes fields.

This was because it was an empty string and not `nil` in these cases. Fixed here: [02fae9c](https://github.com/SaraAlert/SaraAlert/pull/721/commits/02fae9c2b9e1f2a60521a990d41d4d775ea0147d#diff-cdf69598826e24a451843fb600155d000e343d9d9ad213121948b258135c9f99R81)

> 7. When updating the Number of Entries to display in the table, the warning in the screenshot below appears (same thing happens when updating the Number of Entries in the Symptom Assessments table).

So this warning was actually being thrown by any component using the CustomTable. It's because the entries select dropdown returns a string which we then need to parse as an integer. Fortunately, it didn't actually affect functionality because the backend can parse it fine, is just a React prop warning. Fixed here: [02fae9c](https://github.com/SaraAlert/SaraAlert/pull/721/commits/02fae9c2b9e1f2a60521a990d41d4d775ea0147d#diff-36bdccbc821e45bb9795ad5d5f73dbf2f7030b1837d7590d14dccb54d8cd3eb1R148)

kylietmo commented:
@hackrm Blank dose number option now shows up first in modal: a6e3459

holmesie commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+
+  handleProductNameChange = data => {
+    this.setState({ product_name: data.value }, () => {
+      this.checkIfValid();
```
This feels awkward; I'd propose `checkIfValid` does not change state, but instead just reports if things are valid, and set `isValid` at the same time as the first setState, rather than using the callback like this.

timwongj replied:
what about just adding:
```
const isValid = this.state.group_name && this.state.product_name;
```
to the beginning of `render()`
and calling `!isValid` instead of `!this.state.isValid` on line 221?

sgober replied:
you could even just add `this.state.group_name && this.state.product_name` to determine the disabled field

kylietmo replied:
Yeah I originally was thinking the logic might involve more fields and overcomplicated it. Updated here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/7b5ec1bd59c6c8c33fdb9331c5cf35933767c972#diff-182e13a82e02db14eb5222d1eb289b14639dc17acd927608ff7e8b032cfb9af8R68

holmesie commented on app/javascript/components/patient/vaccines/VaccineTable.js:
```diff
@@ -0,0 +1,421 @@
+   * Formats values in the timestamp column to be human readable
+   * @param {Object} data - Data about the cell this filter is called on.
+   */
+  formatTimestamp(data) {
```
It's probably worth centralizing this as a helper somewhere.

kylietmo replied:
I've honestly been dying to do this for awhile. I actually made a [ticket](https://tracker.codev.mitre.org/browse/SARAALERT-654) for it on the backlog in August (lol), just haven't had a chance to actually get a PR up for it. The reason I didn't do it here is because this PR is already quite large and I assumed I'd want to pull this function out from all the other files it was in as well. If we're fine with the additional file changes, I'm okay with adding it and getting it over and done with - or I can just pull this backlog ticket into next sprint.

kylietmo replied:
Update: @DPC15038  will be working on this ticket so whatever work that goes in first can handle this change.

holmesie commented on app/models/vaccine.rb:
```diff
@@ -0,0 +1,57 @@
+  belongs_to :patient
+
+  VACCINE_STANDARDS = Rails.configuration.vaccine_standards.freeze
+  MAX_DOSE_NUMBER = 6
```
Worth putting in a config?

kylietmo replied:
You mean like the vaccine information already is? We could, though this number wasn't necessarily based on maintained info from CDC but rather determined internally. Would be pretty easy to add as another config though

kylietmo commented on app/helpers/vaccine_query_helper.rb:
```diff
@@ -0,0 +1,92 @@
+
+    if sort_direction.present? && !%w[asc desc].include?(sort_direction)
+      error_message = "Unable to sort column in specified direction in request: '#{sort_direction}'"
+      raise StandardError, error_message
```
Yeah I actually created a custom error class at first, just like we already do [here](https://github.com/SaraAlert/SaraAlert/blob/master/app/helpers/patient_query_helper.rb#L587), but given all I wanted to make "custom" about it was insert a special message, one of the linters didn't like it since it was considered a "useless" class since it functioned exactly the same as `StandardError`. I'm happy to go that route again if we think it's worth it.

kylietmo commented on app/javascript/components/admin/AdminTable.js:
```diff
@@ -410,7 +410,7 @@ class AdminTable extends React.Component {
    */
   handleEntriesChange = event => {
-    const value = event.target.value;
+    const value = parseInt(event.target.value);
```
Yep - fixes a bug that Becky brought up here that I realized affected all components using the CustomTable.
See last part of this post: https://github.com/SaraAlert/SaraAlert/pull/721#issuecomment-785329798

timwongj commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+                <DateInput
+                  id="administration-date"
+                  date={this.state.administration_date}
+                  minDate={'1900-01-01'}
```
is this `minDate` this far back because of much older vaccines for other types of diseases?

kylietmo replied:
Yep! This and the max date were decided upon with the PO team (and that was the reasoning).

timwongj commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+                  id="administration-date"
+                  date={this.state.administration_date}
+                  minDate={'1900-01-01'}
+                  maxDate={moment()
```
are we allowing 30 days in the future for scheduled vaccinations?

kylietmo replied:
Yep!

timwongj replied:
I'm guessing that we are not longer allowing future scheduled vaccinations now based on the `maxDate` being set to the current date?

timwongj commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+   * Data is valid and can be saved if there is both a group name and product name selected.
+   */
+  checkIfValid = () => {
+    const isValid = this.state.group_name && this.state.product_name;
```
should we put an asterisk next to the fields that are required to make it more immediately obvious to the users that they are required like we do in other places in the app?

sgober replied:
I agree with this, I was initially confused when testing the disabling of the submit button.  I also noticed we didn't do this with the Close Contact modal either, but there's more either or logic with that one

kylietmo replied:
Yep, absolutely - meant to do this and it slipped my mind. Good catch - will add soon.

kylietmo replied:
Added here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/7b5ec1bd59c6c8c33fdb9331c5cf35933767c972#diff-182e13a82e02db14eb5222d1eb289b14639dc17acd927608ff7e8b032cfb9af8R81

timwongj commented on app/models/history.rb:
```diff
@@ -147,6 +149,14 @@ def self.report_reminder(patient: nil, created_by: 'Sara Alert System', comment:
+    create_history(patient, created_by, HISTORY_TYPES[:vaccination], comment)
+  end
+
+  def self.vaccination_edit(patient: nil, created_by: 'Sara Alert System', comment: 'User edited a vacciation on the monitoree record.')
```
thoughts on making the language for comments between different data types consistent? either by adding more detail to the history messages for the other data types like you did here or making the vaccination messages the same format as the existing history messages?

I'm also thinking that it wouldn't hurt to try to keep the history messages more concise if possible since they probably add up a lot in terms of space in production

kylietmo replied:
Agreed they should be consistent. Updated here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/1af7e86f6a26f91f7b2b6563d00d493acf16da44#diff-08422754c6b565c51a7e094488199ec370790128736cd382aa486715d3126452R154

timwongj commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+                      : ''
+                  }
+                  value={this.state.dose_number ? { label: this.state.dose_number, value: this.state.dose_number } : ''}
+                  options={this.props.dose_number_options?.map(option => {
```
would it make sense to do some kind of validation with dose number in that dose #2 for example cannot come before dose #1 based on the administration date?

kylietmo replied:
This was also discussed in part of the requirement gathering meetings we had, and we decided against it for now.

timwongj commented on app/javascript/components/patient/vaccines/VaccineTable.js:
```diff
@@ -0,0 +1,421 @@
+        </Card>
+        {(this.state.showAddVaccineModal || this.state.showEditVaccineModal) && (
+          <VaccineModal
+            title={this.state.showAddVaccineModal ? 'Add New Vaccination' : 'Edit Vaccination'}
```
should there be consistent language between the 'Add New Vaccine' button and the 'Add New Vaccination' modal title?

kylietmo replied:
Yes, updated here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/7b5ec1bd59c6c8c33fdb9331c5cf35933767c972#diff-36bdccbc821e45bb9795ad5d5f73dbf2f7030b1837d7590d14dccb54d8cd3eb1R350

timwongj commented on app/helpers/vaccine_query_helper.rb:
```diff
@@ -0,0 +1,92 @@
+  end
+
+  # Queries vaccines by ID or type based on search text.
+  def search(vaccines, search)
```
is it too expensive to search by notes?

kylietmo replied:
I had asked about what fields users may want to search by for this and the PO team agreed this functionality likely won't even be used. Given that, we felt it was okay to limit it to a few fields at the moment.

timwongj reviewed (commented):
Looking good so far!

sgober commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+
+  handleAdministrationDateChange = newDate => {
+    this.setState({ administration_date: newDate }, () => {
+      this.checkIfValid();
```
Do we need the checkIfValid here if this field isn't required?

kylietmo replied:
This code has been removed (updated this logic to be more concise).

sgober commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+
+  handleDoseNumberChange = data => {
+    this.setState({ dose_number: data.value }, () => {
+      this.checkIfValid();
```
same comment as above

kylietmo replied:
This code has been removed (updated this logic to be more concise).

sgober commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+  handleNotesChange = event => {
+    const val = event.target.value;
+    this.setState({ notes: val }, () => {
+      this.checkIfValid();
```
same comment as above

kylietmo replied:
This code has been removed (updated this logic to be more concise).

sgober commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+                <Select
+                  inputId="group-name-select"
+                  name="group_name"
+                  defaultValue={
```
Can we break this label/value logic out into its own `getFormattedOptions` function and then find the proper values from that?  See how it's done in the advanced filter, I think it's a lot cleaner and easier to understand

kylietmo replied:
Yeah great idea - done here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/1af7e86f6a26f91f7b2b6563d00d493acf16da44#diff-182e13a82e02db14eb5222d1eb289b14639dc17acd927608ff7e8b032cfb9af8R68
I did this for the `options` attribute, not `defaultValue` since that wasn't always the same pattern. I believe this should resolve all your other comments as well, if not let me know.

sgober commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+                  inputId="product-name-select"
+                  name="product_name"
+                  defaultValue={
+                    this.props.currentVaccineData?.product_name
```
Same for this one

sgober commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,244 @@
+                  inputId="dose-number-select"
+                  name="dose_number"
+                  defaultValue={
+                    this.props.currentVaccineData?.dose_number
```
here as well

sgober reviewed (changes requested):
looking good so far

kylietmo commented on app/models/vaccine.rb:
```diff
@@ -0,0 +1,57 @@
+  MAX_DOSE_NUMBER = 6
+
+  # Additional allowed options for a selected product name aside from the configured official options
+  ADDITIONAL_PRODUCT_NAME_OPTIONS = ['Unknown'].freeze
```
Nope, just Unknown for now.

kylietmo commented on app/models/vaccine.rb:
```diff
@@ -0,0 +1,57 @@
+    }
+  }, presence: { message: 'is required' }
+
+  validates :administration_date, date: true
```
You mean in addition to the date validation it's already doing with the DateValidator, add additional checks on mix/max allowed dates?

timwongj replied:
yep!

kylietmo replied:
Yeah I think that makes sense! I'll add that soon.

kylietmo replied:
Added here @timwongj with some tests as well: https://github.com/SaraAlert/SaraAlert/pull/721/commits/1384f4987e5043fd6c47ee2dce919b948ddc599a

kylietmo commented on app/models/history.rb:
```diff
@@ -147,6 +149,14 @@ def self.report_reminder(patient: nil, created_by: 'Sara Alert System', comment:
     create_history(patient, created_by, HISTORY_TYPES[:report_reminder], comment) unless patient&.preferred_contact_method.nil?
   end
 
+  def self.vaccination(patient: nil, created_by: 'Sara Alert System', comment: 'User added a new vacciation to the monitoree record..')
```
Updated here (as well as another typo): https://github.com/SaraAlert/SaraAlert/pull/721/commits/7b5ec1bd59c6c8c33fdb9331c5cf35933767c972#diff-08422754c6b565c51a7e094488199ec370790128736cd382aa486715d3126452R154

kylietmo commented on test/helpers/vaccine_query_helper_test.rb:
```diff
@@ -0,0 +1,568 @@
+
+  test 'paginate: paginates data based on entries and page' do
+    patient = create(:patient)
+    create(:vaccine, patient: patient)
```
Good call - updated here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/7b5ec1bd59c6c8c33fdb9331c5cf35933767c972#diff-0590548206af8a397cfbb935c41504a4ff5bb70f52d4a4e0acf1bb0cf39e394aR550

kylietmo commented on app/controllers/vaccines_controller.rb:
```diff
@@ -0,0 +1,134 @@
+  before_action :authenticate_user!
+
+  def index
+    redirect_to(root_url) and return unless current_user&.can_view_patient_vaccines?
```
Updated here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/7b5ec1bd59c6c8c33fdb9331c5cf35933767c972#diff-bb55d00f383ddce4b3bf4e2718700168cea42d18736c5017bf567d46a408b15fR10

kylietmo commented:
@timwongj @sgober @holmesie All comments should have been addressed. Let me know if I missed anything.

timwongj commented on app/controllers/vaccines_controller.rb:
```diff
@@ -0,0 +1,134 @@
+
+    # Validate params and handle errors if invalid
+    begin
+      data = validate_table_query(params)
```
can we rename this to `validate_vaccines_query` since the one for patients is `validate_patients_query`?

kylietmo replied:
Yep, done here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/66e9a6d482854054a97a2f0ac8c1ec4c0c6af3ca#diff-cdf69598826e24a451843fb600155d000e343d9d9ad213121948b258135c9f99R6

timwongj commented on app/controllers/vaccines_controller.rb:
```diff
@@ -0,0 +1,134 @@
+    # Verify user has access to patient, patient exists, and the patient has vaccines
+    patient = current_user.get_patient(data[:patient_id])
+    vaccines = patient&.vaccines
+    redirect_to(root_url) && return if patient.nil? || vaccines.blank?
```
Is `vaccines.blank?` going to ever redirect just because the patient doesn't have any vaccinations? That doesn't seem to be the case when I load a patient page without any vaccinations but when I do `Patient.first.vaccines.blank?`, it returns `true`. This probably isn't a problem, but I'm just trying to understand what's going on here

kylietmo replied:
This line is hit when the vaccines are blank. It's not actually redirecting the user ultimately because the request made to get these vaccinations is made asynchronously from a React component which doesn't handle the 302 redirects, so it is just ignored.

timwongj replied:
got it, thanks for the explanation!

timwongj commented on lib/tasks/demo.rake:
```diff
@@ -660,6 +664,42 @@ desc 'Backup the database'
+    patient_ids = existing_patients.pluck(:id).sample(existing_patients.count * rand(15..25) / 100)
+    patient_ids.each_with_index do |patient_id, index|
+      printf("\rGenerating vaccine #{index+1} of #{patient_ids.length}...")
+      changes_ts = create_fake_timestamp(today, today)
```
can we rename this to `vaccine_ts` for consistency

kylietmo replied:
Done: https://github.com/SaraAlert/SaraAlert/pull/721/commits/66e9a6d482854054a97a2f0ac8c1ec4c0c6af3ca#diff-9bfc38ef71b77112eb91f77e35668d17dfea639c79e19de03900b201ad44b8efR674

timwongj commented on app/models/vaccine.rb:
```diff
@@ -0,0 +1,66 @@
+  belongs_to :patient
+
+  VACCINE_STANDARDS = Rails.configuration.vaccine_standards.freeze
+  MAX_DOSE_NUMBER = 2
```
should this number be defined in the config based on the `group_name` and `product_name`? since the J&J vaccine only requires 1 dose and future vaccines might require more/less doses?

kylietmo replied:
Given you and @holmesie both brought up moving this to the config, I went ahead and added a "num_doses" for each vaccine in the config and calculated `MAX_DOSE_NUMBER` based on it: https://github.com/SaraAlert/SaraAlert/pull/721/commits/66e9a6d482854054a97a2f0ac8c1ec4c0c6af3ca#diff-69f2cba2092662388d22d16e62c30957e4c13b866314226ed2ad2580ccf764b6R10

timwongj replied:
thanks Kylie, would it make sense to only allow the `MAX_DOSE_NUMBER` based on the `group_name` and `product_name` that way a user can't enter in dose_number 2 for J&J or would that just make things too complicated?

kylietmo replied:
Yeah when we were discussing all this we considered a few options for further validation (checking administration dates base on dose number, verifying you don't have records for the same vaccine with the same dose number, etc) but we ultimately landed on not trying to make Sara Alert a vaccine management system and keeping it simple for now, at least for this first iteration

timwongj replied:
sounds good!

timwongj commented on app/javascript/components/patient/vaccines/VaccineModal.js:
```diff
@@ -0,0 +1,232 @@
+
+  /**
+   */
+  checkIfValid = () => {};
```
I think this can be removed

kylietmo replied:
Good catch, removed here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/66e9a6d482854054a97a2f0ac8c1ec4c0c6af3ca#diff-182e13a82e02db14eb5222d1eb289b14639dc17acd927608ff7e8b032cfb9af8L29

sgober commented on app/javascript/tests/patient/vaccines/VaccineModal.test.js:
```diff
@@ -0,0 +1,22 @@
+});
+
+describe('VaccineModal', () => {
+  it('TODO', () => {
```
Are you planning on writing these before you go?  If not, I'd say remove the files and create a JIRA to write them specifically for these two components

kylietmo replied:
That was originally the plan but time has slipped away so yes I made a ticket to address. I'm honestly okay with keeping them for now as skeleton files unless you're strongly opposed. Ticket is [here](https://tracker.codev.mitre.org/browse/SARAALERT-1379).

sgober replied:
In the past, we haven't kept any skeleton files to reduce confusion.  Since there isn't anything substantial in here (new mocks, any rendering tests, etc.), I'd say just remove it because who knows when we'll get to that ticket

sgober reviewed (commented):
Looks good!  Just have one question (see below) then will approve

holmesie reviewed (approved):
Awesome work!
